# Preprocesamiento de datos para train y test

Empiezo cargando las librerias necesarias e importando los archivos .csv

In [1]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

train_values = pd.read_csv("Training set values.csv")
train_labels = pd.read_csv("Training set labels.csv")
test_values = pd.read_csv("Test set values.csv")

Hago una primera visual para ver los datos de los 2 dataframes (train y test)

In [2]:
pd.set_option('display.max_columns', None)
train_values.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  \
0  34.938093  -9.856322                  none            0   
1  34.698766  -2.147466              Zahanati            0   
2  37.460664  -3.821329           Kwa Mahundi            0   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4  31.130847  -1.825359               Shuleni            0   

                     basin  subvillage   region  region_code  district_code  \
0               Lake Nyasa    Mnyusi B   Iringa           11              5   
1            Lake Victoria     Nyamara     Mara           20              2   
2                  Pangani     Majengo  Manyara           21              4   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   
4            Lake Victoria  Kyanyamisa   Kagera           18              1   

         lga        ward  population public_meeting              recorded_by  \
0     Ludewa    Mundindi         109           True  GeoData Consultants Ltd   
1  Serengeti       Natta         280            NaN  GeoData Consultants Ltd   
2  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd   
3   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd   
4    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd   

  scheme_management                  scheme_name permit  construction_year  \
0               VWC                        Roman  False               1999   
1             Other                          NaN   True               2010   
2               VWC  Nyumba ya mungu pipe scheme   True               2009   
3               VWC                          NaN   True               1986   
4               NaN                          NaN   True                  0   

  extraction_type extraction_type_group extraction_type_class management  \
0         gravity               gravity               gravity        vwc   
1         gravity               gravity               gravity        wug   
2         gravity               gravity               gravity        vwc   
3     submersible           submersible           submersible        vwc   
4         gravity               gravity               gravity      other   

  management_group         payment payment_type water_quality quality_group  \
0       user-group    pay annually     annually          soft          good   
1       user-group       never pay    never pay          soft          good   
2       user-group  pay per bucket   per bucket          soft          good   
3       user-group       never pay    never pay          soft          good   
4            other       never pay    never pay          soft          good   

       quantity quantity_group                source           source_type  \
0        enough         enough                spring                spring   
1  insufficient   insufficient  rainwater harvesting  rainwater harvesting   
2        enough         enough                   dam                   dam   
3           dry            dry           machine dbh              borehole   
4      seasonal       seasonal  rainwater harvesting  rainwater harvesting   

  source_class              waterpoint_type waterpoint_type_group  
0  groundwater           communal standpipe    communal standpipe  
1      surface           communal standpipe    communal standpipe  
2      surface  communal standpipe multiple    communal standpipe  
3  groundwater  communal standpipe multiple    communal standpipe 

In [3]:
pd.set_option('display.max_columns', None)
test_values.head()

id  amount_tsh date_recorded                  funder  gps_height  \
0  50785         0.0    2013-02-04                    Dmdd        1996   
1  51630         0.0    2013-02-04  Government Of Tanzania        1569   
2  17168         0.0    2013-02-01                     NaN        1567   
3  45559         0.0    2013-01-22              Finn Water         267   
4  49871       500.0    2013-03-27                  Bruder        1260   

    installer  longitude   latitude                 wpt_name  num_private  \
0        DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
1         DWE  36.656709  -3.309214                  Kimnyak            0   
2         NaN  34.767863  -5.004344           Puma Secondary            0   
3  FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
4      BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   

                     basin  subvillage   region  region_code  district_code  \
0                 Internal      Magoma  Manyara           21              3   
1                  Pangani     Kimnyak   Arusha            2              2   
2                 Internal       Msatu  Singida           13              2   
3  Ruvuma / Southern Coast  Kipindimbi    Lindi           80             43   
4  Ruvuma / Southern Coast     Losonga   Ruvuma           10              3   

             lga          ward  population public_meeting  \
0          Mbulu        Bashay         321           True   
1   Arusha Rural      Kimnyaki         300           True   
2  Singida Rural          Puma         500           True   
3         Liwale       Mkutano         250            NaN   
4         Mbinga  Mbinga Urban          60            NaN   

               recorded_by scheme_management     scheme_name permit  \
0  GeoData Consultants Ltd        Parastatal             NaN   True   
1  GeoData Consultants Ltd               VWC  TPRI pipe line   True   
2  GeoData Consultants Ltd               VWC               P    NaN   
3  GeoData Consultants Ltd               VWC             NaN   True   
4  GeoData Consultants Ltd       Water Board          BRUDER   True   

   construction_year extraction_type extraction_type_group  \
0               2012           other                 other   
1               2000         gravity               gravity   
2               2010           other                 other   
3               1987           other                 other   
4               2000         gravity               gravity   

  extraction_type_class   management management_group      payment  \
0                 other   parastatal       parastatal    never pay   
1               gravity          vwc       user-group    never pay   
2                 other          vwc       user-group    never pay   
3                 other          vwc       user-group      unknown   
4               gravity  water board       user-group  pay monthly   

  payment_type water_quality quality_group      quantity quantity_group  \
0    never pay          soft          good      seasonal       seasonal   
1    never pay          soft          good  insufficient   insufficient   
2    never pay          soft          good  insufficient   insufficient   
3      unknown          soft          good           dry            dry   
4      monthly          soft          good        enough         enough   

                 source           source_type source_class  \
0  rainwater harvesting  rainwater harvesting      surface   
1                spring                spring  groundwater   
2  rainwater harvesting  rainwater harvesting      surface   
3          shallow well          shallow well  groundwater   
4                spring                spring  groundwater   

      waterpoint_type waterpoint_type_group  
0               other                 other  
1  communal standpipe    communal standpipe  
2               other                 other  
3               other      

Veo las variables numéricas y veo que tanto population, num_private y construction_year tienen una gran cantidad de '0'.

In [4]:
train_values.describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04   
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00   
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00   
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01   
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00   
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00   
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00   
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   

        num_private   region_code  district_code    population  \
count  59400.000000  59400.000000   59400.000000  59400.000000   
mean       0.474141     15.297003       5.629747    179.909983   
std       12.236230     17.587406       9.633649    471.482176   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     25.000000   
75%        0.000000     17.000000       5.000000    215.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year  
count       59400.000000  
mean         1300.652475  
std           951.620547  
min             0.000000  
25%             0.000000  
50%          1986.000000  
75%          2004.000000  
max          2013.000000

Veo las variables categóricas y se pueden ver cierta similitud entre los distintos grupos de variables como 'extraction', 'quantity', 'source', 'waterpoint' o 'payment'

In [5]:
train_values.describe(include=['object', 'category'])


date_recorded                  funder installer wpt_name  \
count          59400                   55763     55745    59398   
unique           356                    1896      2145    37399   
top       2011-03-15  Government Of Tanzania       DWE     none   
freq             572                    9084     17402     3563   

                basin subvillage  region     lga   ward public_meeting  \
count           59400      59029   59400   59400  59400          56066   
unique              9      19287      21     125   2092              2   
top     Lake Victoria   Madukani  Iringa  Njombe  Igosi           True   
freq            10248        508    5294    2503    307          51011   

                    recorded_by scheme_management scheme_name permit  \
count                     59400             55522       30590  56344   
unique                        1                11        2695      2   
top     GeoData Consultants Ltd               VWC           K   True   
freq                      59400             36793         682  38852   

       extraction_type extraction_type_group extraction_type_class management  \
count            59400                 59400                 59400      59400   
unique              18                    13                     7         12   
top            gravity               gravity               gravity        vwc   
freq             26780                 26780                 26780      40507   

       management_group    payment payment_type water_quality quality_group  \
count             59400      59400        59400         59400         59400   
unique                5          7            7             8             6   
top          user-group  never pay    never pay          soft          good   
freq              52490      25348        25348         50818         50818   

       quantity quantity_group  source source_type source_class  \
count     59400          59400   59400       59400        59400   
unique        5              5      10           7            3   
top      enough         enough  spring      spring  groundwater   
freq      33186          33186   17021       17021        45794   

           waterpoint_type waterpoint_type_group  
count                59400                 59400  
unique                   7                     6  
top     communal standpipe    communal standpipe  
freq                 28522                 34625

In [6]:
train_values.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55763 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59398 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

Miro los missings de cada variable (veo que 'scheme_name tiene la mitad) y miro también los valores únicos de cada variable para tener una mejor idea de los datos

In [7]:
df_nulls = pd.DataFrame(
    [(col, train_values[col].isnull().sum()) for col in train_values.columns if train_values[col].isnull().sum() > 0],
    columns=["Variable", "Missings"]
)
df_nulls = df_nulls.sort_values(by="Missings", ascending=False).reset_index(drop=True)
print(df_nulls)

            Variable  Missings
0        scheme_name     28810
1  scheme_management      3878
2          installer      3655
3             funder      3637
4     public_meeting      3334
5             permit      3056
6         subvillage       371
7           wpt_name         2


In [8]:
unique_counts = train_values.nunique().sort_values(ascending=True)
print(unique_counts)


recorded_by                  1
public_meeting               2
permit                       2
source_class                 3
quantity                     5
quantity_group               5
management_group             5
quality_group                6
waterpoint_type_group        6
payment                      7
source_type                  7
waterpoint_type              7
payment_type                 7
extraction_type_class        7
water_quality                8
basin                        9
source                      10
scheme_management           11
management                  12
extraction_type_group       13
extraction_type             18
district_code               20
region                      21
region_code                 27
construction_year           55
num_private                 65
amount_tsh                  98
lga                        125
date_recorded              356
population                1049
funder                    1896
ward                      2092
installe

Renombro los valores de las variables para que todas sean minúsculas, pero comiencen con mayúsculas, en ambos dataframes, ya que he observado que algunas variables tienen el mismo nombre pero una en minúscula y otras en mayúsculas.

In [9]:
train_values = train_values.applymap(lambda x: x.capitalize() if isinstance(x, str) else x)
test_values =test_values.applymap(lambda x: x.capitalize() if isinstance(x, str) else x)

C:\Users\josem\AppData\Local\Temp\ipykernel_11060\2937159763.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  train_values = train_values.applymap(lambda x: x.capitalize() if isinstance(x, str) else x)
C:\Users\josem\AppData\Local\Temp\ipykernel_11060\2937159763.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  test_values =test_values.applymap(lambda x: x.capitalize() if isinstance(x, str) else x)


El siguiente paso es unir los datos de entrenamiento con las etiquedas mediante un .merge teniendo como punto de unión los 'id'. También guardo los 'id' que me servirán más tarde para generar el archivo de submission a evaluar.

In [10]:
train_data = pd.merge(train_values, train_labels, on='id')
test_data = test_values.copy()
test_ids = test_data['id']

Veo que la variable Recorded_by sólo tiene 1 valor en train_values por lo que no me da nada de información y la borro. También observo que en public_meeting los False vienen enmascarados como NaN, por lo que hago este cambio

In [11]:
train_data['public_meeting'].value_counts()


public_meeting
True     51011
False     5055
Name: count, dtype: int64

In [12]:
true_count = train_data['public_meeting'].value_counts().get(True, 0)
nan_count = train_data['public_meeting'].isna().sum()

print(f"True: {true_count}")
print(f"NaN: {nan_count}")

True: 51011
NaN: 3334


In [13]:
train_data = train_data.drop(columns=['recorded_by'])
test_data = test_data.drop(columns=['recorded_by'])

train_data['public_meeting'] = train_data['public_meeting'].fillna(False)
test_data['public_meeting'] = test_data['public_meeting'].fillna(False)

C:\Users\josem\AppData\Local\Temp\ipykernel_11060\1914292792.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  train_data['public_meeting'] = train_data['public_meeting'].fillna(False)
C:\Users\josem\AppData\Local\Temp\ipykernel_11060\1914292792.py:5: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  test_data['public_meeting'] = test_data['public_meeting'].fillna(False)


Después de verificar similitudes entre distintas parejas o grupos de variables, veo que region_code tiene 27 valores distintos, mientras que region tiene 21. Las comparo y veo que algunas regiones tienen varios códigos por lo que la información de region_code está contenida en region, así que la borro

In [14]:
train_data['region'].nunique()

21

In [15]:
train_data['region_code'].nunique()

27

In [16]:
# Identificar las regiones que tienen más de un código
region_code_duplicates = train_data.groupby('region')['region_code'].nunique()
regions_with_multiple_codes = region_code_duplicates[region_code_duplicates > 1]

regions_with_multiple_codes

train_data = train_data.drop(columns=['region_code'])
test_data = test_data.drop(columns=['region_code'])

Después de muchas pruebas, veo que lo mejor para el preprocesado de datos es quedarnos con las demás variables (después haremos otra limpieza más general), imputar los valores faltantes por la moda (para categóricas) y la mediana (para numéricas)

In [17]:
categorical_columns = train_data.select_dtypes(include='object').columns.drop('status_group')
for column in categorical_columns:
    train_data[column] = train_data[column].fillna(train_data[column].mode()[0])
    test_data[column] = test_data[column].fillna(test_data[column].mode()[0])

numerical_columns = train_data.select_dtypes(include=['float64', 'int64']).columns
for column in numerical_columns:
    train_data[column] = train_data[column].fillna(train_data[column].median())
    test_data[column] = test_data[column].fillna(test_data[column].median())

C:\Users\josem\AppData\Local\Temp\ipykernel_11060\1442608120.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  train_data[column] = train_data[column].fillna(train_data[column].mode()[0])
C:\Users\josem\AppData\Local\Temp\ipykernel_11060\1442608120.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  test_data[column] = test_data[column].fillna(test_data[column].mode()[0])


Elimino la variable 'id' que no me aporta información y ya los tengo guardados en test_ids. En cuanto a la variable 'scheme_name', la borro también ya que tiene 28810 missings (la mitad de los datos aproximadamente)

In [18]:
train_data.drop(columns=['scheme_name', 'id'], inplace=True, errors='ignore')
test_data.drop(columns=['scheme_name', 'id'], inplace=True, errors='ignore')

Paso las fechas a datetime mediante una función y borro la variable original. Con los años de esta fecha creo una variable nueva que incluya la antigüedad (recorded - construction)

In [19]:
def procesar_fechas(data):
    data['date_recorded'] = pd.to_datetime(data['date_recorded'])
    data['recorded_year'] = data['date_recorded'].dt.year
    data['recorded_month'] = data['date_recorded'].dt.month
    data['recorded_day'] = data['date_recorded'].dt.day
    data.drop(columns=['date_recorded'], inplace=True)
    return data

train_data = procesar_fechas(train_data)
test_data = procesar_fechas(test_data)

In [20]:
median_year = train_data.loc[train_data['construction_year'] > 0, 'construction_year'].median()
train_data['construction_year'] = train_data['construction_year'].replace(0, median_year)
test_data['construction_year'] = test_data['construction_year'].replace(0, median_year)
train_data['waterpoint_age'] = train_data['recorded_year'] - train_data['construction_year']
test_data['waterpoint_age'] = test_data['recorded_year'] - test_data['construction_year']

En cuando a la geografía de cada bomba, las agrupo en 10 clusters mediante KMeans basándome en la latitud y longitud. Creo la variable geo_cluster que incluya esta información

In [21]:
coords = pd.concat([train_data[['latitude', 'longitude']], test_data[['latitude', 'longitude']]])
kmeans = KMeans(n_clusters=10, random_state=42).fit(coords)
train_data['geo_cluster'] = kmeans.predict(train_data[['latitude', 'longitude']])
test_data['geo_cluster'] = kmeans.predict(test_data[['latitude', 'longitude']])

El siguiente paso es separar la variable objetivo de los datos de entrenamiento y eliminar las columnas categóricas que tengan demasiadas categorías (alta cardinalidad), para poder convertirlas a dummies y que no generen demasiadas columnas.

In [22]:
y = train_data['status_group']
train_data.drop(columns=['status_group'], inplace=True)

cat_cols = train_data.select_dtypes(include='object').columns
high_card_cols = [col for col in cat_cols if train_data[col].nunique() > 100]
train_data.drop(columns=high_card_cols, inplace=True)
test_data.drop(columns=high_card_cols, inplace=True, errors='ignore')

train_data = pd.get_dummies(train_data, drop_first=True)
test_data = pd.get_dummies(test_data, drop_first=True)

Alineo las columnas del test para que tenga las mismas columnas del train, rellenando con ceros si faltan algunas. Se escalan las variables numéricas para mejorar el rendimiento del modelo

In [23]:
test_data = test_data.reindex(columns=train_data.columns, fill_value=0)

In [24]:
X = train_data
scaler = StandardScaler()
X = scaler.fit_transform(X)
test_data = scaler.transform(test_data)

# Una vez finalizado el procesamiento de los datos, pasamos al modelado predictivo

Optimizamos el modelo mediante RandomizedSearchCV en el que incluimos varios parámetros a explorar y después de probar muchas combinaciones cruzadas nos elige la mejor, para a continuación predecir los resultados mediante RandomForest (el modelo que me ha dado mayor accuracy).

###### Tarda un poquito en ejecutar, uso el n_jobs=-1 para que aproveche toda la CPU de mi ordenador (en mi caso tarda unos 40 segundos)

In [ ]:
model = RandomForestClassifier(random_state=42, n_jobs=-1)
param_dist = {
    'n_estimators': [100, 200, 300],
    'max_depth': [20, 30, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
random_search = RandomizedSearchCV(model, param_dist, n_iter=10, cv=3, random_state=42, n_jobs=-1)
random_search.fit(X, y)

In [ ]:
best_model = random_search.best_estimator_

y_test_pred = best_model.predict(test_data)

### Por último generamos el archivo a evaluar que tenga el formato especificado en el enunciado y lo probamos en la pagina web del concurso

In [ ]:
submission = pd.DataFrame({
    'id': test_ids,
    'status_group': y_test_pred
})
submission.to_csv('Submission_Jose_Manuel_Jimenez_Serrano.csv', index=False)

### El score obtenido mediante este procedimiento es de 0,8185

## Pruebas realizadas en otros notebooks, durante el procesado de los datos, pero que me han reducido el score:
- He probado a borrar payment_type ya que contiene, prácticamente, los mismos valores que payment pero me ha reducido el score 0,0017. Lo mismo me ocurrió con quantity y quantity_group.

- He probado a ir borrando variables que parecían contenerse en otras (como quedarme con 'extraction_type' y borrar las otras 2 variables de extractions, por ejemplo). Esto me ha ocurrido con las demás variables que empiezan por el mismo nombre (suelen estar seguidas en la visualización del .head).

- He probado a sustituir los meses y días (despues de pasar la fecha a años, meses y dias) por las estaciones de Tanzania (agrupándolas como 1 Primavera, 2 Verano, 3 Otoño y 4 Invierno).

- Otra prueba ha sido la de ir, mediante funciones lambda, sustituyendo variables numéricas con muchas categorías por 10 categorías como mucho, viendo previamente los rangos de cada variable, ya que cada una tiene un rango y distribución diferentes. Pero me daban scores entre los 0,77 y los 0,79, por lo que al final decidí quedarme con todas aquellas que tengan menos de 100 categorías, pero con la distribución por defecto.

- También he probado a borrar 'num_private', ya que 58643 valores de los 59400 eran ceros, por lo que pensé que tampoco nos daría información importante, pero al borrarla reduce el score otro 0,0017.

- Con 'amount_tsh' me pasó igual que con 'num_private', tiene 41639 ceros, que seguramente sean datos que se han añadido por falta de información del proveedor, pero que al borrarla me reduce el score.

- También he comparado entre 'funder' e 'installer' porque muchos valores se repetían y quizás quedándome con una de las 2 variables obtendría un score similar, pero no fue el caso.

- Otra prueba fue la de borrar 'quality_group' ya que 'water_quality' contiene valores muy parecidos pero ésta es más precisa en cuanto a la calidad del agua.

- Dentro de 'subvillage' había valores raros que pensé que eran missings (M, 1, I, etc), aunque tenían una frecuencia de unos 100, pero que no podía encajar en ninguna otra variable (por falta de información) y tampoco podía borrar porque perdía score.

- He probado también a unir dentro de 'source' la categoría Other con Unknown, ya que en teoría me está dando la misma información (Otra diferente o desconocida deberían dar la misma información).

- Otra prueba que he hecho ha sido exportar los datos de train y test, y que con pycaret, y también de forma manual con código, me prueben distintos modelos con este procesamiento de datos y el que mayor acuraccy me ha dado siempre, ha sido el RandomForest, seguido de SVC.